source/extensions/filters/http/jwt_authn/verifier.cc

```diff
@@ -0,0 +1,306 @@
+#include "extensions/filters/http/jwt_authn/verifier.h"
+
+#include "common/common/enum_to_int.h"
+#include "common/http/message_impl.h"
+#include "common/http/utility.h"
+
+using ::envoy::config::filter::http::jwt_authn::v2alpha::JwtProvider;
+using ::envoy::config::filter::http::jwt_authn::v2alpha::JwtRequirement;
+using ::envoy::config::filter::http::jwt_authn::v2alpha::JwtRequirementAndList;
+using ::envoy::config::filter::http::jwt_authn::v2alpha::JwtRequirementOrList;
+using ::Envoy::Http::LowerCaseString;
+using ::google::jwt_verify::Status;
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace JwtAuthn {
+namespace {
+
+class BaseChildVerifierContexImpl : public VerifierContext {
+public:
+  BaseChildVerifierContexImpl(VerifierContext& parent) : parent_(parent), is_completed_(false) {}
+
+  virtual void complete(const Status& status) override {
+    if (is_completed_) {
+      return;
+    }
+    is_completed_ = true;
+    parent_.complete(status);
+  }
+
+  bool isCompleted() const override { return is_completed_; }
+
+  Http::HeaderMap& headers() const override { return parent_.headers(); }
+
+protected:
+  VerifierContext& parent_;
+  bool is_completed_;
+};
+
+class AuthenticatorVerifierContextImpl : public BaseChildVerifierContexImpl,
+                                         public Authenticator::Callbacks {
+public:
+  AuthenticatorVerifierContextImpl(VerifierContext& parent, AuthenticatorPtr&& auth,
+                                   bool can_bypass)
+      : BaseChildVerifierContexImpl(parent), auth_(std::move(auth)), can_bypass_(can_bypass) {
+    auth_->verify(headers(), this);
```

Comment:
I have very hard time to understand this logic.   You hide "verify" action in side the context.  

Context is only for holding dynamic data.   Not action.   

Please refactor it that  "verify" such action is called in the Verifier interface verify() method call.

Context is only for data.  

I have hard time to read there are so many ContextImpl.  It should have only one ContextImpl.

Reply:
Context is the callbacks for authenticators. I thought this is what you meant when you said callbacks should belong in the context too.

Reply:
I mean,  context is storing callback as data.   Not actual using it.